### PR TITLE
aws-efs-csi-driver helm workaround

### PR DESCRIPTION
**Pull request by NormJohnIV**

### CODE REVIEW

**Review by enderm — CHANGES_REQUESTED** [review]
got this:

TASK [efs-provisioner : Deploy aws-efs-csi-driver] *****************************
fatal: [localhost]: FAILED! => {"changed": false, "command": "/usr/local/bin/helm --kubeconfig /tmp/ansible.r1ka4tbc/.kube --namespace=kube-system --version=1.0.0 upgrade -i --reset-values --wait aws-efs-csi-driver https://github.com/kubernetes-sigs/aws-efs-csi-driver/releases/download/v1.0.0/helm-chart.tgz", "msg": "Failure when executing Helm command. Exited 1.\nstdout: Release \"aws-efs-csi-driver\" does not exist. Installing it now.\n\nstderr: Error: rendered manifests contain a resource that already exists. Unable to continue with install: CSIDriver \"efs.csi.aws.com\" in namespace \"\" exists and cannot be imported into the current release: invalid ownership metadata; label validation error: missing key \"app.kubernetes.io/managed-by\": must be set to \"Helm\"; annotation validation error: missing key \"meta.helm.sh/release-name\": must be set to \"aws-efs-csi-driver\"; annotation validation error: missing key \"meta.helm.sh/release-namespace\": must be set to \"kube-system\"\n", "stderr": "Error: rendered manifests contain a resource that already exists. Unable to continue with install: CSIDriver \"efs.csi.aws.com\" in namespace \"\" exists and cannot be imported into the current release: invalid ownership metadata; label validation error: missing key \"app.kubernetes.io/managed-by\": must be set to \"Helm\"; annotation validation error: missing key \"meta.helm.sh/release-name\": must be set to \"aws-efs-csi-driver\"; annotation validation error: missing key \"meta.helm.sh/release-namespace\": must be set to \"kube-system\"\n", "stderr_lines": ["Error: rendered manifests contain a resource that already exists. Unable to continue with install: CSIDriver \"efs.csi.aws.com\" in namespace \"\" exists and cannot be imported into the current release: invalid ownership metadata; label validation error: missing key \"app.kubernetes.io/managed-by\": must be set to \"Helm\"; annotation validation error: missing key \"meta.helm.sh/release-name\": must be set to \"aws-efs-csi-driver\"; annotation validation error: missing key \"meta.helm.sh/release-namespace\": must be set to \"kube-system\""], "stdout": "Release \"aws-efs-csi-driver\" does not exist. Installing it now.\n", "stdout_lines": ["Release \"aws-efs-csi-driver\" does not exist. Installing it now."]}

**Comment by NormJohnIV** [comment]
@enderm That error states you already had a csidriver installed that is missing the helm annotations. I believe this is an artifact from the previous install attempts with the fault helm manifest. This can be cleared by running "kubectl delete csidrivers.storage.k8s.io efs.csi.aws.com". Also, delete secret "sh.helm.release.v1.aws-efs-csi-driver.v1" if it exists in the kube-system namespace

File: roles/efs-provisioner/tasks/main.yml [hunk 1]
```diff
@@ -6,28 +6,17 @@
   with_items:
     - name: "stable" 
       url: "https://charts.helm.sh/stable"
-    - name: "aws-efs-csi-driver" 
-      url: "https://kubernetes-sigs.github.io/aws-efs-csi-driver/"
+    # - name: "aws-efs-csi-driver"
```
Comment (thpang) [inline roles/efs-provisioner/tasks/main.yml:9]: These could be removed. Any reason to keep them as comments?

**Comment by thpang** [comment]
Except for the comment above, code looks good.

**Comment by thpang** [comment]
@enderm re-test based on Norm's information above.

**Comment by enderm** [comment]
hm, things failed for me on a brand-new cluster.
After deleting the driver, the helm install works. Also, after doing a "baseline,uninstall" I can successfully install the efs-csi  driver.
BUT, against a brand-new cluster, I still get the error.
Looks like the csi driver is being installed by default with eks, at least when using the module that we use in viay4-iac-aws
Here is what I have:

$ kubectl get csidrivers.storage.k8s.io -o yaml
apiVersion: v1
items:
- apiVersion: storage.k8s.io/v1
  kind: CSIDriver
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"storage.k8s.io/v1","kind":"CSIDriver","metadata":{"annotations":{},"name":"efs.csi.aws.com"},"spec":{"attachRequired":false}}
    creationTimestamp: "2020-12-08T15:19:57Z"
    managedFields:
    - apiVersion: storage.k8s.io/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            .: {}
            f:kubectl.kubernetes.io/last-applied-configuration: {}
        f:spec:
          f:attachRequired: {}
          f:podInfoOnMount: {}
          f:volumeLifecycleModes:
            .: {}
            v:"Persistent": {}
      manager: kubectl
      operation: Update
      time: "2020-12-08T15:19:57Z"
    name: efs.csi.aws.com
    resourceVersion: "171"
    selfLink: /apis/storage.k8s.io/v1/csidrivers/efs.csi.aws.com
    uid: dfab5f2b-31a2-4157-a8c4-c1cc64ec305b
  spec:
    attachRequired: false
    podInfoOnMount: false
    volumeLifecycleModes:
    - Persistent
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""

**Comment by enderm** [comment]
correction: ansible 'uninstall, baseline' does actually NOT remove the pre-installed csi driver.

**Comment by NormJohnIV** [comment]
From further testing, we do not actually need the aws efs csi driver. So, it has been removed. You should be able to deploy fine without it now.

**Comment by enderm** [comment]
sigh, no luck with this latest one either. Against a fresh cluster:
TASK [efs-provisioner : Add helm repos] ****************************************
changed: [localhost]
Wednesday 09 December 2020  01:38:07 +0000 (0:00:00.734)       0:00:04.704 ****

TASK [Deploy efs-provisioner] **************************************************
fatal: [localhost]: FAILED! => {"changed": false, "command": "/usr/local/bin/helm --kubeconfig /tmp/ansible.0nz3ad0w/.kube --namespace=efs-provisioner --version=0.12.0 show chart stable/efs-provisioner", "msg": "Failure when executing Helm command. Exited 1.\nstdout: \nstderr: Error: failed to download \"stable/efs-provisioner\" at version \"0.12.0\" (hint: running `helm repo update` may help)\n", "stderr": "Error: failed to download \"stable/efs-provisioner\" at version \"0.12.0\" (hint: running `helm repo update` may help)\n", "stderr_lines": ["Error: failed to download \"stable/efs-provisioner\" at version \"0.12.0\" (hint: running `helm repo update` may help)"], "stdout": "", "stdout_lines": []}

**Comment by thpang** [comment]
@enderm are you running from within the docker container? If so, can you jump into the container and try the helm command by hand and see if the helm repo might be the problem. Thx.

**Comment by enderm** [comment]
yes, I'm using the container. 
Running the helm command manually gets the same result:
$ /usr/local/bin/helm --kubeconfig /tmp/ansible.uy2bcx18/.kube --namespace=efs-provisioner --version=0.12.0 show chart stable/efs-provisioner
Error: failed to download "stable/efs-provisioner" at version "0.12.0" (hint: running `helm repo update` may help)

and: $ helm repo update
Error: no repositories found. You must add one before updating

**Comment by NormJohnIV** [comment]
The repo is added only when the tool runs. By default, the container has no repos. Do some searching, we see that the efs provisioner from the previous helm stable repo is now deprecated. The only active efs-provisioner is provided by banzai cloud at version 0.0.2. Therefore, I will switch over to using it.

Screen Shot 2020-12-10 at 4 34 39 PM

**Comment by NormJohnIV** [comment]
The banzai cloud version is quite outdated and not a "proper" helm chart. So, The better option would be to package the tgz to the latest stable version in this repo.

**Comment by NormJohnIV** [comment]
EFS helm chart has been added to the efs-provisioner role's files.

**Comment by enderm** [comment]
TASK [Deploy efs-provisioner] **************************************************
fatal: [localhost]: FAILED! => {"changed": false, "command": "/usr/local/bin/helm --kubeconfig /tmp/ansible.okd7poq7/.kube repo update", "msg": "Failure when executing Helm command. Exited 1.\nstdout: \nstderr: Error: no repositories found. You must add one before updating\n", "stderr": "Error: no repositories found. You must add one before updating\n", "stderr_lines": ["Error: no repositories found. You must add one before updating"], "stdout": "", "stdout_lines": []}

**Comment by NormJohnIV** [comment]
I disabled update_repo_cache on the efs-provisioner since there is no helm repo in the docker container when it runs and the chart is local anyways.

**Comment by enderm** [comment]
looks good now.

**Review by enderm — APPROVED** [review]

**Pull request merged**